src/radical/pilot/agent/lm/jsrun.py

```diff
@@ -66,25 +71,55 @@ def _create_resource_set_file(self, slots, uid, sandbox):
 
         uid     : unit ID (string)
         sandbox : unit sandbox (string)
+        mpi     : MPI or not (bool, default: False)
 
         """
 
-        rs_id  = 0
-        rs_str = ''
-
-        for node in slots['nodes']:
-
-            cores = ' '.join([str(core_set[0]) for core_set
-                                               in  node['core_map']])
-            gpus  = ' '.join([str(gpu_set[0])  for gpu_set
-                                               in  node['gpu_map']])
-
-            rs_str           += 'RS %d: {'  % rs_id
-            rs_str           += ' host: %d' % node['uid']
-            if cores: rs_str += ' cpu: %s'  % cores
-            if gpus : rs_str += ' gpu: %s'  % gpus
-            rs_str           += ' }\n'
-            rs_id            += 1
+        rs_str = 'cpu_index_using: physical\n'
+        if mpi and gpu:
+            rank = 0
+            for node in slots['nodes']:
+                for cores_set, gpu_set in zip(node['core_map'],
+                                              node['gpu_map']):
+
+                    cores = ','.join(str(core) for core in cores_set)
+                    gpus  = ','.join(str(gpu) for gpu in gpu_set)
+
+                    rs_str += 'rank: %d: {'  % rank
+                    rs_str += ' host: %d;'  % node['uid']
+                    rs_str += ' cpu: {%s}'  % cores
+                    rs_str += '; gpu: {%s}' % gpus
+                    rs_str += '}\n'
+                    rank   += 1
+        elif mpi and not gpu:
+            rank = 0
+            for node in slots['nodes']:
+                for cores_set in node['core_map']:
+
+                    cores = ','.join(str(core) for core in cores_set)
+
+                    rs_str += 'rank: %d: {'  % rank
+                    rs_str += ' host: %d;'  % node['uid']
+                    rs_str += ' cpu: {%s}'  % cores
+                    rs_str += '}\n'
+                    rank   += 1
+        else:
+            for node in slots['nodes']:
+                cores_str = list()
+                gpus_str = list()
+                for core_set in node['core_map']:
+                    cores_str = ['{' + ','.join(str(core) for core
+                                      in core_set) + '}'] 
+                for gpu_set in node['gpu_map']:
+                    gpus_str += ['{' + ','.join(str(gpu) for gpu
+                                      in gpu_set) + '}']
+                cores = ','.join(str(core) for core in cores_str)
+                gpus = ','.join(str(gpu) for gpu in gpus_str)
+                rs_str           += '1: {'
+                rs_str           += ' host: %d;' % node['uid']
+                if cores: rs_str += ' cpu: %s' % cores
+                if gpus : rs_str += '; gpu: %s' % gpus
+                rs_str           += '}\n'
```

Comment:
I do not understand what you mean. Can you give an example or propose a change?

Reply:
Sure. See example resource layouts. 'rank' is only useful to specify rank-core binding when they are out of order. See example 9 of https://www.ibm.com/support/knowledgecenter/en/SSWRJV_10.1.0/jsm/10.2/base/erf_format.html. Otherwise with and without 'rank' are equivalent.

With 'rank':
```bash
# MPI task
bash-4.2$ cat erf_1.rs 
cpu_index_using: physical
rank: 0: {host: 1; cpu: {0,1,2,3}}
rank: 1: {host: 1; cpu: {6,7,8,9}}

bash-4.2$ jsrun --erf_input erf_1.rs /ccs/proj/csc190/Hello_jsrun/jsrun_layout
########################################################################

*** MPI Ranks: 2, OpenMP Threads: 4, GPUs per Resource Set: 0 ***
========================================================================
MPI Rank 000 of 002 on HWThread 000 of Node a30n01, OMP_threadID 0 of 4
MPI Rank 000 of 002 on HWThread 001 of Node a30n01, OMP_threadID 1 of 4
MPI Rank 000 of 002 on HWThread 003 of Node a30n01, OMP_threadID 2 of 4
MPI Rank 000 of 002 on HWThread 002 of Node a30n01, OMP_threadID 3 of 4
MPI Rank 001 of 002 on HWThread 007 of Node a30n01, OMP_threadID 1 of 4
MPI Rank 001 of 002 on HWThread 006 of Node a30n01, OMP_threadID 0 of 4
MPI Rank 001 of 002 on HWThread 008 of Node a30n01, OMP_threadID 3 of 4
MPI Rank 001 of 002 on HWThread 009 of Node a30n01, OMP_threadID 2 of 4

# Non-MPI task (i.e. task with 1 process)
bash-4.2$ cat erf_2.rs 
rank: 0: {host: 1; cpu: {0,1,2,3}}

bash-4.2$ jsrun --erf_input erf_2.rs /ccs/proj/csc190/Hello_jsrun/jsrun_layout
########################################################################

*** MPI Ranks: 1, OpenMP Threads: 4, GPUs per Resource Set: 0 ***
========================================================================
MPI Rank 000 of 001 on HWThread 000 of Node a30n01, OMP_threadID 1 of 4
MPI Rank 000 of 001 on HWThread 003 of Node a30n01, OMP_threadID 2 of 4
MPI Rank 000 of 001 on HWThread 001 of Node a30n01, OMP_threadID 0 of 4
MPI Rank 000 of 001 on HWThread 002 of Node a30n01, OMP_threadID 3 of 4
```

Without rank:
```bash
# MPI task
bash-4.2$ cat erf_3.rs 
1: {host: 1; cpu: {0,1,2,3}}
1: {host: 1; cpu: {6,7,8,9}}

bash-4.2$ jsrun --erf_input erf_3.rs /ccs/proj/csc190/Hello_jsrun/jsrun_layout
MPI Rank 001 of 002 on HWThread 007 of Node a30n01, OMP_threadID 1 of 4
MPI Rank 001 of 002 on HWThread 009 of Node a30n01, OMP_threadID 2 of 4
MPI Rank 001 of 002 on HWThread 008 of Node a30n01, OMP_threadID 3 of 4
MPI Rank 001 of 002 on HWThread 006 of Node a30n01, OMP_threadID 0 of 4
########################################################################

*** MPI Ranks: 2, OpenMP Threads: 4, GPUs per Resource Set: 0 ***
========================================================================
MPI Rank 000 of 002 on HWThread 001 of Node a30n01, OMP_threadID 3 of 4
MPI Rank 000 of 002 on HWThread 000 of Node a30n01, OMP_threadID 0 of 4
MPI Rank 000 of 002 on HWThread 002 of Node a30n01, OMP_threadID 1 of 4
MPI Rank 000 of 002 on HWThread 003 of Node a30n01, OMP_threadID 2 of 4

# Non-MPI task (i.e. task with 1 process)
bash-4.2$ cat erf_4.rs 
cpu_index_using: physical
1: {host: 1; cpu: {0,1,2,3}}

bash-4.2$ jsrun --erf_input erf_4.rs /ccs/proj/csc190/Hello_jsrun/jsrun_layout
########################################################################

*** MPI Ranks: 1, OpenMP Threads: 4, GPUs per Resource Set: 0 ***
========================================================================
MPI Rank 000 of 001 on HWThread 000 of Node a30n01, OMP_threadID 0 of 4
MPI Rank 000 of 001 on HWThread 003 of Node a30n01, OMP_threadID 1 of 4
MPI Rank 000 of 001 on HWThread 002 of Node a30n01, OMP_threadID 2 of 4
MPI Rank 000 of 001 on HWThread 001 of Node a30n01, OMP_threadID 3 of 4
```

IMO we don't really require both 'rank' and without 'rank' modes. Its enough to keep one.
That said, I don't think we need 'rank' since we won't be doing out of order rank-core binding since we need application knowledge for that. My suggestion would to be go with no specification about the rank (i.e. 'without rank' in my above example which is equivalent to the 'else' case you have in the implementation).

Just for completion what I mean by out of order rank-core binding:
We will always assign resources to rank 0, then to rank 1 and so on. 'rank' is only useful
when we want to assign resources to rank j before rank i where j>i. (example 9 of https://www.ibm.com/support/knowledgecenter/en/SSWRJV_10.1.0/jsm/10.2/base/erf_format.html)

Let me know if you see something I am missing.

Reply:
I do not think that `Specific ordering` requires out of order necessarily. I prefer the specificity of the ranks because it leaves no room for errors in interpretation. I can remove them if it causes so much trouble.

Reply:
No it doesn't. Purpose of rank is to specify an order. We are choosing it to be a particular order (ascending order), that's all. I do think we are over-specifying it. My suggestion would be to not use 'rank'. 

Either way I would recommend to use one approach for mpi and non-mpi tasks.

Reply:
So @andre-merzky any opinions on this? It will affect also the way test for #1822 change with the ERF format!

Reply:
@andre-merzky Any opinion? I am happy with keeping as general as possible

Reply:
Based on last evenings Slack heated discussion, we are staying with the explicit rank style files.